cdap-common/src/main/java/co/cask/cdap/gateway/handlers/SecureHandler.java

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.gateway.handlers;
+
+import co.cask.cdap.common.authorization.SubjectIds;
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.common.http.SecurityRequestContext;
+import co.cask.common.authorization.IdentifiableObject;
+import co.cask.common.authorization.ObjectId;
+import co.cask.common.authorization.Permission;
+import co.cask.common.authorization.UnauthorizedException;
+import co.cask.common.authorization.client.AuthorizationClient;
+import co.cask.http.HttpResponder;
+import com.google.common.base.Function;
+import com.google.common.base.Objects;
+import com.google.common.base.Optional;
+import com.google.common.base.Splitter;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Multimap;
+import com.google.inject.Inject;
+import org.jboss.netty.handler.codec.http.HttpResponseStatus;
+
+/**
+ * Provides methods which provide security to netty-http handlers.
+ */
+public class SecureHandler {
```

Comment:
I think it's wrong to do it with the current way. When a user is authenticated, the actual handler (e.g. `ProgramLifecycleHandler`) has access to the security context, knowing who is calling, hence can already produce a list that is visible to that user only, isn't it? It's none of security filter business on mutating it since it's application data, isn't it?